### ISSUE TITLE
Unable to find absent links using AbsentLink

### ISSUE THREAD

**Issue by stellarspot** [opencog/atomspace#2017]
I have a simple example where are a cat and a stone and the cat is inherited from animal.
My aim is to find all atoms which are not animals and I would expect that it should return stone but it does not:
```
(use-modules (opencog) (opencog query) (opencog exec))

(define cat (Concept "cat"))
(define stone (Concept "stone"))
(define animal (Concept "animal"))

(Inheritance cat animal)

(define pattern
 (Bind
  (Variable "$X")
  (Absent
   (Inheritance
    (Variable "$X")
    animal))
  (Variable "$X")))

(display
 (cog-execute! pattern))
```

**Comment by rTreutlein** [comment]
Your pattern should look like this:

(define pattern
 (Bind
  (TypedVariableLink
	(Variable "$X")
	(TypeNode "ConceptNode"))
  (AndLink
	  (Absent
	    (Inheritance
		(Variable "$X")
		animal))
          (Present
	   (Variable "$X")))
  (Variable "$X")))

If you only have the Absent Link it only looks for all Inheritance Links of things that are animals and checks that they don't exist. So, I also added (Variable "$X") to the query so now it a looks for all things.
Which would result in an infinite loop as the query itself could be the answer? So you also need to specify the type of the Variable. This then leads to the result.

Basically, if you only have an AbsentLink in your query it will never return anything as it will only look for things specified inside of the AbsentLink which of course all get's discarded because of the AbsentLink.

**Comment by linas** [comment]
Specifying the type of the variable also helps limit the search, slightly. The more things you can do to limit the size of the ssearch space, the faster it will run; the `AbsentLink` is always done last, after all other search filters have been applied.

**Comment by linas** [comment]
Closing, I assume this is no longer an issue!?
